cmd/helm/install.go

```diff
@@ -201,6 +201,7 @@ func addInstallFlags(cmd *cobra.Command, f *pflag.FlagSet, client *action.Instal
 	f.StringToStringVarP(&client.Labels, "labels", "l", nil, "Labels that would be added to release metadata. Should be divided by comma.")
 	f.BoolVar(&client.EnableDNS, "enable-dns", false, "enable DNS lookups when rendering templates")
 	f.BoolVar(&client.HideNotes, "hide-notes", false, "if set, do not show notes in install output. Does not affect presence in chart metadata")
+	f.BoolVar(&client.ThreeWayMergeForCustomResources, "three-way-merge-for-custom-resources", false, "also considers the state of custom resources and custom resource definitions in the cluster when adopting resources")
```

Comment:
Isn't this redundant for install? Even when adopting resources, there isn't going to be a previous version (only the new and real versions). So no three-way merge will be possible (only two-way)

Unless I'm mistaken?

Reply:
No, it is not redundant.

Adoption of resources end up being performed in Client.Update (helm/pkg/kube/client.go), even if the install command was used. For that to work, it effectively sets the original and target versions of a resource to the same content, regardless of the merge strategy used. Later, when the patch is to be created, it checks for the types to differentiate between native and unstructured resources and fetches the current version of the resources from the cluster.

If a two-way merge is used for patch creation, only the original and target versions are compared, never the current version of the cluster. This leads the patch creation code to return an empty patch, so that the unstructured resource is not changed.

If a three-way-merge is used, the current version is also considered and that enables updates of unstructured resources. 

Additionally but potentially less relevant, if `TakeOwnership` (only in SDK, not a CLI flag) is used, which is supposed to work regardless of the ownership of resources, the result looks more like a failure (although Helm doesn't currently see it as such), since the resources also aren't touched, but that means they end up in a state not owned by helm. Having that three-way-merge flag in the CLI for adoption with TakeOwnership is not necessary, as TakeOwnership cannot currently be enabled from the CLI, but it could prevent developers missing to add it if TakeOwnership is made available to the CLI eventually.

Reply:
fwiw, the CLI should gain `--take-ownership` support: https://github.com/helm/helm/pull/13439